### DISCUSSION TITLE
What is the guideline or procedure for improvements of one of the existing implementations?

### DISCUSSION THREAD

**Discussion by fvbakel** [PlummersSoftwareLLC/Primes#222] [Q&A] [answered]
What is the guideline or procedure for (small) performance improvements of one of the existing implementations?

Different cases:
1. There is a small performance improvement possible in the Python solution 2, it was made by someone else.
2. There is a small performance improvement possible in the Tcl solution 1, that was made by my self.
3. A small improvement in the algorithm can be applied to multiple implementations. 

So what should be done in these cases? 

Considerations:

- Raise an issue for each solution that can be improved
- Update the existing code and submit a new pull request?
- Implement a different script in the same solution directory, with a different name and make sure both are run by the Dockerfile?
- Implement a new solution directory, with only minor changes to the original?
- Does it matter if it is an improvement of one of my "own" or from somebody else?

Please advice.

**Comment by rbergen** [comment, marked as answer]
- You can always open PRs for changes to your own solutions, however small. We'll merge these quickly, provided that they don't break the solution and stay in line with its documented characteristics (algorithm, faithfulness, bits per prime). 

- You can open a PR for a solution authored by someone else, if the change you're making is not too big. You can ask for a review by the author by mentioning them in the PR description; I've recently done this myself in #219. We will then allow the original author to respond (and voice any objections they may have). If it is a small and obvious improvement and the original author doesn't respond, we are likely to still merge the PR after some time. At his specific request, we will _not_ apply changes that change the performance of solutions written by @davepl.

- If your suggested change to someone else's solution is more significant, to the point that it changes key characteristics of that solution, it makes sense to create a new solution based on the original one, and clearly credit the author of the original solution in the source file(s) and README.md. The new solution can then be submitted in accordance with the contributing guidelines.

- If you plan on making conceptually identical improvements to multiple solutions, it makes sense to start a discussion in the category "Ideas", first. That way you can get feedback on what you intend to do before you spend the time and energy on making the change.

- Blanket improvements to multiple solutions can be suggested in a PR, after the discussion I just mentioned took place, or not. Please explain what you're trying to achieve with the change in the PR description; objective improvements to the "fairness" of performance comparisons across solutions are generally speaking a good reason. We will review the merits of the bulk change on a case-by-case basis.

**Reply by fvbakel** [reply]
Thank you, perfect clear to me.
